{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d30b3288-b3f0-4ff2-a57d-802ac46bf0ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "import allel\n",
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37b4133e-e59c-473b-9fe4-b74978e76338",
   "metadata": {},
   "source": [
    "## JPT population \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "727bed58-7590-48dc-8061-e236bd103960",
   "metadata": {},
   "source": [
    "#### JPT data with allel package ###"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "53d8d3da-6f56-4d5f-b562-7c46ee091520",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div class=\"allel allel-DisplayAs2D\"><span>&lt;GenotypeArray shape=(1059517, 104, 2) dtype=int8&gt;</span><table><thead><tr><th></th><th style=\"text-align: center\">0</th><th style=\"text-align: center\">1</th><th style=\"text-align: center\">2</th><th style=\"text-align: center\">3</th><th style=\"text-align: center\">4</th><th style=\"text-align: center\">...</th><th style=\"text-align: center\">99</th><th style=\"text-align: center\">100</th><th style=\"text-align: center\">101</th><th style=\"text-align: center\">102</th><th style=\"text-align: center\">103</th></tr></thead><tbody><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">0</th><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">...</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">1</th><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">...</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">2</th><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">...</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">...</th><td style=\"text-align: center\" colspan=\"12\">...</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">1059514</th><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">...</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">1059515</th><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">...</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">1059516</th><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">...</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "<GenotypeArray shape=(1059517, 104, 2) dtype=int8>\n",
       "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
       "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
       "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
       "...\n",
       "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
       "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0\n",
       "0/0 0/0 0/0 0/0 0/0 ... 0/0 0/0 0/0 0/0 0/0"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "JPTdata = allel.read_vcf('JPT_SNPs_only.vcf', fields='*') ## Read the VCF file using allel \n",
    "\n",
    "JPTgenotypes=JPTdata['calldata/GT'] ## extract the genotypes array \n",
    "\n",
    "JPTarray=allel.GenotypeArray(JPTgenotypes) \n",
    "JPTarray## convert the genotypes data frame into a genotype array \n",
    "\n",
    "JPTgenotypes.n_samples ## count the total number of samples that will be used later to calculate genotypes frequencies "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2286f25c-bded-404a-ac3c-33a2264f9061",
   "metadata": {},
   "source": [
    "#### JPT data extracted with vcf tools ###"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f543d225-6bbe-4c2b-a4fb-e1db2e9e6fbc",
   "metadata": {},
   "source": [
    "## vcf tools command "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5cfa92a8-3943-485a-b224-5553bc6df132",
   "metadata": {},
   "outputs": [],
   "source": [
    "##  \"--vcf JPT_SNPs.vcf --hardy --out JPT.hwe \""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "3bf1605e-560e-4cff-9b19-f983e2cfc0ef",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>OBS(HOM1/HET/HOM2)</th>\n",
       "      <th>E(HOM1/HET/HOM2)</th>\n",
       "      <th>ChiSq_HWE</th>\n",
       "      <th>P_HWE</th>\n",
       "      <th>P_HET_DEFICIT</th>\n",
       "      <th>P_HET_EXCESS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22</td>\n",
       "      <td>16050075</td>\n",
       "      <td>104/0/0</td>\n",
       "      <td>104.00/0.00/0.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22</td>\n",
       "      <td>16050115</td>\n",
       "      <td>104/0/0</td>\n",
       "      <td>104.00/0.00/0.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>22</td>\n",
       "      <td>16050213</td>\n",
       "      <td>104/0/0</td>\n",
       "      <td>104.00/0.00/0.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22</td>\n",
       "      <td>16050319</td>\n",
       "      <td>104/0/0</td>\n",
       "      <td>104.00/0.00/0.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22</td>\n",
       "      <td>16050527</td>\n",
       "      <td>104/0/0</td>\n",
       "      <td>104.00/0.00/0.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055449</th>\n",
       "      <td>22</td>\n",
       "      <td>51241342</td>\n",
       "      <td>104/0/0</td>\n",
       "      <td>104.00/0.00/0.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055450</th>\n",
       "      <td>22</td>\n",
       "      <td>51241386</td>\n",
       "      <td>102/2/0</td>\n",
       "      <td>102.01/1.98/0.01</td>\n",
       "      <td>0.009803</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.995169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055451</th>\n",
       "      <td>22</td>\n",
       "      <td>51244163</td>\n",
       "      <td>104/0/0</td>\n",
       "      <td>104.00/0.00/0.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055452</th>\n",
       "      <td>22</td>\n",
       "      <td>51244205</td>\n",
       "      <td>104/0/0</td>\n",
       "      <td>104.00/0.00/0.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055453</th>\n",
       "      <td>22</td>\n",
       "      <td>51244237</td>\n",
       "      <td>104/0/0</td>\n",
       "      <td>104.00/0.00/0.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1055454 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         CHR       POS OBS(HOM1/HET/HOM2)  E(HOM1/HET/HOM2)  ChiSq_HWE  P_HWE  \\\n",
       "0         22  16050075            104/0/0  104.00/0.00/0.00        NaN    1.0   \n",
       "1         22  16050115            104/0/0  104.00/0.00/0.00        NaN    1.0   \n",
       "2         22  16050213            104/0/0  104.00/0.00/0.00        NaN    1.0   \n",
       "3         22  16050319            104/0/0  104.00/0.00/0.00        NaN    1.0   \n",
       "4         22  16050527            104/0/0  104.00/0.00/0.00        NaN    1.0   \n",
       "...      ...       ...                ...               ...        ...    ...   \n",
       "1055449   22  51241342            104/0/0  104.00/0.00/0.00        NaN    1.0   \n",
       "1055450   22  51241386            102/2/0  102.01/1.98/0.01   0.009803    1.0   \n",
       "1055451   22  51244163            104/0/0  104.00/0.00/0.00        NaN    1.0   \n",
       "1055452   22  51244205            104/0/0  104.00/0.00/0.00        NaN    1.0   \n",
       "1055453   22  51244237            104/0/0  104.00/0.00/0.00        NaN    1.0   \n",
       "\n",
       "         P_HET_DEFICIT  P_HET_EXCESS  \n",
       "0                  1.0      1.000000  \n",
       "1                  1.0      1.000000  \n",
       "2                  1.0      1.000000  \n",
       "3                  1.0      1.000000  \n",
       "4                  1.0      1.000000  \n",
       "...                ...           ...  \n",
       "1055449            1.0      1.000000  \n",
       "1055450            1.0      0.995169  \n",
       "1055451            1.0      1.000000  \n",
       "1055452            1.0      1.000000  \n",
       "1055453            1.0      1.000000  \n",
       "\n",
       "[1055454 rows x 8 columns]"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "jpt_data_frame = pd.read_csv ('jpt.hwe',delimiter=\"\\t\") ## Read the hwe file extrcated from vcf tools \n",
    "jpt_data_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "ea888c08-a0be-429a-9ee0-ee4d2abc0e41",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OBS(HOM1/HET/HOM2)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>104/0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>104/0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>104/0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>104/0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>104/0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055449</th>\n",
       "      <td>104/0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055450</th>\n",
       "      <td>102/2/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055451</th>\n",
       "      <td>104/0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055452</th>\n",
       "      <td>104/0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055453</th>\n",
       "      <td>104/0/0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1055454 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        OBS(HOM1/HET/HOM2)\n",
       "0                  104/0/0\n",
       "1                  104/0/0\n",
       "2                  104/0/0\n",
       "3                  104/0/0\n",
       "4                  104/0/0\n",
       "...                    ...\n",
       "1055449            104/0/0\n",
       "1055450            102/2/0\n",
       "1055451            104/0/0\n",
       "1055452            104/0/0\n",
       "1055453            104/0/0\n",
       "\n",
       "[1055454 rows x 1 columns]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "keep_col = ['OBS(HOM1/HET/HOM2)'] ## extract only the columns with the genotypes \n",
    "jpt_data_frame=jpt_data_frame[keep_col] ## save it as a dataframe \n",
    "jpt_data_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "69fe1239-0255-4839-abf2-b7b60099d209",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "JPT=jpt_data_frame['OBS(HOM1/HET/HOM2)'].str.split('/', expand=True)## split the data frame to get seperate column for each genotype \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "d177a44a-5bc4-4ac5-b297-cab2260ce322",
   "metadata": {},
   "outputs": [],
   "source": [
    "JPT=JPT.rename(columns={JPT.columns[0]:\"JPT_0|0 \"})## rename the columns with the country code and the corresponding genotype symbols \n",
    "JPT=JPT.rename(columns={JPT.columns[1]:\"JPT_0|1 \"})\n",
    "JPT=JPT.rename(columns={JPT.columns[2]:\"JPT_1|1 \"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "e1c74d9d-1487-4dc2-8ea2-6c4a015c0309",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>JPT_0|0</th>\n",
       "      <th>JPT_0|1</th>\n",
       "      <th>JPT_1|1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>104</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>104</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>104</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>104</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>104</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055449</th>\n",
       "      <td>104</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055450</th>\n",
       "      <td>102</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055451</th>\n",
       "      <td>104</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055452</th>\n",
       "      <td>104</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055453</th>\n",
       "      <td>104</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1055454 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         JPT_0|0   JPT_0|1   JPT_1|1 \n",
       "0             104         0         0\n",
       "1             104         0         0\n",
       "2             104         0         0\n",
       "3             104         0         0\n",
       "4             104         0         0\n",
       "...           ...       ...       ...\n",
       "1055449       104         0         0\n",
       "1055450       102         2         0\n",
       "1055451       104         0         0\n",
       "1055452       104         0         0\n",
       "1055453       104         0         0\n",
       "\n",
       "[1055454 rows x 3 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "JPT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "b2d50250-569f-426c-9c76-efddab953c73",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>JPT_0|0</th>\n",
       "      <th>JPT_0|1</th>\n",
       "      <th>JPT_1|1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055449</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055450</th>\n",
       "      <td>0.980769</td>\n",
       "      <td>0.019231</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055451</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055452</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055453</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1055454 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         JPT_0|0   JPT_0|1   JPT_1|1 \n",
       "0        1.000000  0.000000       0.0\n",
       "1        1.000000  0.000000       0.0\n",
       "2        1.000000  0.000000       0.0\n",
       "3        1.000000  0.000000       0.0\n",
       "4        1.000000  0.000000       0.0\n",
       "...           ...       ...       ...\n",
       "1055449  1.000000  0.000000       0.0\n",
       "1055450  0.980769  0.019231       0.0\n",
       "1055451  1.000000  0.000000       0.0\n",
       "1055452  1.000000  0.000000       0.0\n",
       "1055453  1.000000  0.000000       0.0\n",
       "\n",
       "[1055454 rows x 3 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "JPT = JPT.astype(int) #convert the date frame date into int \n",
    "JPT.div(104).## divide the genotypes by the number of samples (extracted earlier with allel ) to get the genotype frequencies "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "585313c2-8d6d-41a0-a872-c702d24ad2ac",
   "metadata": {
    "tags": []
   },
   "source": [
    "### GWD population"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b730ae03-bac5-4951-90be-3dd0d8f56206",
   "metadata": {},
   "source": [
    "#### GWD data with allel package ###"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9304d69-ac9b-4669-8d78-2838019db9ca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "113"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GWDdata = allel.read_vcf('GWD_SNPs_only.vcf', fields='*') ## Read the VCF file using allel \n",
    "\n",
    "GWDgenotypes=GWDdata['calldata/GT'] ## extract the genotypes array \n",
    "\n",
    "GWDarray=allel.GenotypeArray(GWDgenotypes) ## convert the genotypes data frame into a genotype array \n",
    "\n",
    "GWDarray.n_samples ## count the total number of samples that will be used later to calculate genotypes frequencies\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62a12898-2089-4bde-a9c3-18e7703f2f8a",
   "metadata": {},
   "source": [
    "#### GWD data extracted with vcf tools ###"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "b83a03a9-9b0a-400f-8b4a-1476bf4b52e3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>OBS(HOM1/HET/HOM2)</th>\n",
       "      <th>E(HOM1/HET/HOM2)</th>\n",
       "      <th>ChiSq_HWE</th>\n",
       "      <th>P_HWE</th>\n",
       "      <th>P_HET_DEFICIT</th>\n",
       "      <th>P_HET_EXCESS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22</td>\n",
       "      <td>16050075</td>\n",
       "      <td>113/0/0</td>\n",
       "      <td>113.00/0.00/0.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22</td>\n",
       "      <td>16050115</td>\n",
       "      <td>110/3/0</td>\n",
       "      <td>110.02/2.96/0.02</td>\n",
       "      <td>0.020451</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.986667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>22</td>\n",
       "      <td>16050213</td>\n",
       "      <td>99/14/0</td>\n",
       "      <td>99.43/13.13/0.43</td>\n",
       "      <td>0.492791</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.651472</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22</td>\n",
       "      <td>16050319</td>\n",
       "      <td>113/0/0</td>\n",
       "      <td>113.00/0.00/0.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22</td>\n",
       "      <td>16050527</td>\n",
       "      <td>113/0/0</td>\n",
       "      <td>113.00/0.00/0.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055449</th>\n",
       "      <td>22</td>\n",
       "      <td>51241342</td>\n",
       "      <td>112/1/0</td>\n",
       "      <td>112.00/1.00/0.00</td>\n",
       "      <td>0.002232</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055450</th>\n",
       "      <td>22</td>\n",
       "      <td>51241386</td>\n",
       "      <td>108/5/0</td>\n",
       "      <td>108.06/4.89/0.06</td>\n",
       "      <td>0.057841</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.955855</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055451</th>\n",
       "      <td>22</td>\n",
       "      <td>51244163</td>\n",
       "      <td>113/0/0</td>\n",
       "      <td>113.00/0.00/0.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055452</th>\n",
       "      <td>22</td>\n",
       "      <td>51244205</td>\n",
       "      <td>113/0/0</td>\n",
       "      <td>113.00/0.00/0.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055453</th>\n",
       "      <td>22</td>\n",
       "      <td>51244237</td>\n",
       "      <td>113/0/0</td>\n",
       "      <td>113.00/0.00/0.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1055454 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         CHR       POS OBS(HOM1/HET/HOM2)  E(HOM1/HET/HOM2)  ChiSq_HWE  P_HWE  \\\n",
       "0         22  16050075            113/0/0  113.00/0.00/0.00        NaN    1.0   \n",
       "1         22  16050115            110/3/0  110.02/2.96/0.02   0.020451    1.0   \n",
       "2         22  16050213            99/14/0  99.43/13.13/0.43   0.492791    1.0   \n",
       "3         22  16050319            113/0/0  113.00/0.00/0.00        NaN    1.0   \n",
       "4         22  16050527            113/0/0  113.00/0.00/0.00        NaN    1.0   \n",
       "...      ...       ...                ...               ...        ...    ...   \n",
       "1055449   22  51241342            112/1/0  112.00/1.00/0.00   0.002232    1.0   \n",
       "1055450   22  51241386            108/5/0  108.06/4.89/0.06   0.057841    1.0   \n",
       "1055451   22  51244163            113/0/0  113.00/0.00/0.00        NaN    1.0   \n",
       "1055452   22  51244205            113/0/0  113.00/0.00/0.00        NaN    1.0   \n",
       "1055453   22  51244237            113/0/0  113.00/0.00/0.00        NaN    1.0   \n",
       "\n",
       "         P_HET_DEFICIT  P_HET_EXCESS  \n",
       "0                  1.0      1.000000  \n",
       "1                  1.0      0.986667  \n",
       "2                  1.0      0.651472  \n",
       "3                  1.0      1.000000  \n",
       "4                  1.0      1.000000  \n",
       "...                ...           ...  \n",
       "1055449            1.0      1.000000  \n",
       "1055450            1.0      0.955855  \n",
       "1055451            1.0      1.000000  \n",
       "1055452            1.0      1.000000  \n",
       "1055453            1.0      1.000000  \n",
       "\n",
       "[1055454 rows x 8 columns]"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwd_data_frame = pd.read_csv ('GWD.hwe',delimiter=\"\\t\") ## Read the hwe file extrcated from vcf tools \n",
    "gwd_data_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "e61f25d6-1fe0-4e48-8a42-48215c3d1c1f",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OBS(HOM1/HET/HOM2)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>113/0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>110/3/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>99/14/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>113/0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>113/0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055449</th>\n",
       "      <td>112/1/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055450</th>\n",
       "      <td>108/5/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055451</th>\n",
       "      <td>113/0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055452</th>\n",
       "      <td>113/0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055453</th>\n",
       "      <td>113/0/0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1055454 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        OBS(HOM1/HET/HOM2)\n",
       "0                  113/0/0\n",
       "1                  110/3/0\n",
       "2                  99/14/0\n",
       "3                  113/0/0\n",
       "4                  113/0/0\n",
       "...                    ...\n",
       "1055449            112/1/0\n",
       "1055450            108/5/0\n",
       "1055451            113/0/0\n",
       "1055452            113/0/0\n",
       "1055453            113/0/0\n",
       "\n",
       "[1055454 rows x 1 columns]"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "keep_col = ['OBS(HOM1/HET/HOM2)'] ## extract only the columns with the genotypes \n",
    "gwd_data_frame=gwd_data_frame[keep_col] ## save it as a dataframe \n",
    "gwd_data_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "ccbe398e-6aeb-4023-a5f8-fe30c7459d5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "GWD=gwd_data_frame['OBS(HOM1/HET/HOM2)'].str.split('/', expand=True)## split the data frame to get seperate column for each genotype "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "8f5e6b82-baa0-4df1-a926-52ad4683ee4a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "GWD=GWD.rename(columns={GWD.columns[0]:\"GWD_0|0 \"})## rename the columns with the country code and the corresponding genotype symbols \n",
    "GWD=GWD.rename(columns={GWD.columns[1]:\"GWD_0|1 \"})\n",
    "GWD=GWD.rename(columns={GWD.columns[2]:\"GWD_1|1 \"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "32aeb6ae-e74d-486d-b5f7-ddbf3edb4bf8",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GWD_0|0</th>\n",
       "      <th>GWD_0|1</th>\n",
       "      <th>GWD_1|1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.973451</td>\n",
       "      <td>0.026549</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.876106</td>\n",
       "      <td>0.123894</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055449</th>\n",
       "      <td>0.991150</td>\n",
       "      <td>0.008850</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055450</th>\n",
       "      <td>0.955752</td>\n",
       "      <td>0.044248</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055451</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055452</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055453</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1055454 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         GWD_0|0   GWD_0|1   GWD_1|1 \n",
       "0        1.000000  0.000000       0.0\n",
       "1        0.973451  0.026549       0.0\n",
       "2        0.876106  0.123894       0.0\n",
       "3        1.000000  0.000000       0.0\n",
       "4        1.000000  0.000000       0.0\n",
       "...           ...       ...       ...\n",
       "1055449  0.991150  0.008850       0.0\n",
       "1055450  0.955752  0.044248       0.0\n",
       "1055451  1.000000  0.000000       0.0\n",
       "1055452  1.000000  0.000000       0.0\n",
       "1055453  1.000000  0.000000       0.0\n",
       "\n",
       "[1055454 rows x 3 columns]"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GWD = GWD.astype(int) #convert the date frame date into int \n",
    "GWD.div(113)## divide the genotypes by the number of samples (extracted earlier with allel ) to get the genotype frequencies"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0480c9e8-9755-4607-b681-9bba931674ea",
   "metadata": {},
   "source": [
    "## GIH population"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2db2bd7-2609-49d5-ab13-de87a5789bdf",
   "metadata": {},
   "source": [
    "#### GIH data with allel package ###\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "90f8588a-df91-403a-9795-69c1b4a0107f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "103"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GIHdata = allel.read_vcf('GIH_SNPs_only.vcf', fields='*') ## Read the VCF file using allel \n",
    "\n",
    "GIHgenotypes=GIHdata['calldata/GT'] ## extract the genotypes array \n",
    "\n",
    "GIHarray=allel.GenotypeArray(GIHgenotypes) ## convert the genotypes data frame into a genotype array \n",
    "\n",
    "GIHarray.n_samples ## count the total number of samples that will be used later to calculate genotypes frequencies\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "5ad6e233-56bc-4358-a262-c6ee985d428e",
   "metadata": {},
   "outputs": [],
   "source": [
    "gih_data_frame = pd.read_csv ('gih.hwe',delimiter=\"\\t\") ## Read the hwe file extrcated from vcf tools "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "b68e1142-7099-4131-931a-20a98e3dbe95",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OBS(HOM1/HET/HOM2)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>103/0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>103/0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>103/0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>103/0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>103/0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055449</th>\n",
       "      <td>103/0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055450</th>\n",
       "      <td>102/1/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055451</th>\n",
       "      <td>103/0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055452</th>\n",
       "      <td>103/0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055453</th>\n",
       "      <td>100/3/0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1055454 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        OBS(HOM1/HET/HOM2)\n",
       "0                  103/0/0\n",
       "1                  103/0/0\n",
       "2                  103/0/0\n",
       "3                  103/0/0\n",
       "4                  103/0/0\n",
       "...                    ...\n",
       "1055449            103/0/0\n",
       "1055450            102/1/0\n",
       "1055451            103/0/0\n",
       "1055452            103/0/0\n",
       "1055453            100/3/0\n",
       "\n",
       "[1055454 rows x 1 columns]"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "keep_col = ['OBS(HOM1/HET/HOM2)'] ## extract only the columns with the genotypes \n",
    "gih_data_frame=gih_data_frame[keep_col] ## save it as a dataframe \n",
    "gih_data_frame\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "7ad4d259-5acc-4a07-ba6a-404441e1b07b",
   "metadata": {},
   "outputs": [],
   "source": [
    "GIH=gih_data_frame['OBS(HOM1/HET/HOM2)'].str.split('/', expand=True)## split the data frame to get seperate column for each genotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "62a4b4af-ba12-4730-aff8-09d23667f8b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "GIH=GIH.rename(columns={GIH.columns[0]:\"GIH_0|0 \"})## rename the columns with the country code and the corresponding genotype symbols \n",
    "GIH=GIH.rename(columns={GIH.columns[1]:\"GIH_0|1 \"})\n",
    "GIH=GIH.rename(columns={GIH.columns[2]:\"GIH_1|1 \"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "f295ac04-a901-498a-abc7-ca0834eaa0cf",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GIH_0|0</th>\n",
       "      <th>GIH_0|1</th>\n",
       "      <th>GIH_1|1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055449</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055450</th>\n",
       "      <td>0.990291</td>\n",
       "      <td>0.009709</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055451</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055452</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055453</th>\n",
       "      <td>0.970874</td>\n",
       "      <td>0.029126</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1055454 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         GIH_0|0   GIH_0|1   GIH_1|1 \n",
       "0        1.000000  0.000000       0.0\n",
       "1        1.000000  0.000000       0.0\n",
       "2        1.000000  0.000000       0.0\n",
       "3        1.000000  0.000000       0.0\n",
       "4        1.000000  0.000000       0.0\n",
       "...           ...       ...       ...\n",
       "1055449  1.000000  0.000000       0.0\n",
       "1055450  0.990291  0.009709       0.0\n",
       "1055451  1.000000  0.000000       0.0\n",
       "1055452  1.000000  0.000000       0.0\n",
       "1055453  0.970874  0.029126       0.0\n",
       "\n",
       "[1055454 rows x 3 columns]"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GIH = GIH.astype(int) #convert the date frame date into int \n",
    "GIH.div(103)## divide the genotypes by the number of samples (extracted earlier with allel ) to get the genotype frequencies"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "994704bf-c919-41f3-b11e-729d31095e8f",
   "metadata": {},
   "source": [
    "## PUR population"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b486d4ac-4e44-47cc-a5f6-c2fa0ff362e1",
   "metadata": {},
   "source": [
    "#### PUR data with allel package ###\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "5d6da86e-15a0-4242-82c8-5ad1358f3ef3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "104"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PURdata = allel.read_vcf('PUR_SNPs_only.vcf', fields='*') ## Read the VCF file using allel \n",
    "\n",
    "PURgenotypes=PURdata['calldata/GT'] ## extract the genotypes array \n",
    "\n",
    "PURarray=allel.GenotypeArray(PURgenotypes) ## convert the genotypes data frame into a genotype array \n",
    "\n",
    "PURarray.n_samples ## count the total number of samples that will be used later to calculate genotypes frequencies\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "3415a267-7d10-459a-896a-c11544c8da3a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OBS(HOM1/HET/HOM2)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>104/0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>104/0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>104/0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>104/0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>104/0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055449</th>\n",
       "      <td>104/0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055450</th>\n",
       "      <td>102/2/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055451</th>\n",
       "      <td>104/0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055452</th>\n",
       "      <td>104/0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055453</th>\n",
       "      <td>102/2/0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1055454 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        OBS(HOM1/HET/HOM2)\n",
       "0                  104/0/0\n",
       "1                  104/0/0\n",
       "2                  104/0/0\n",
       "3                  104/0/0\n",
       "4                  104/0/0\n",
       "...                    ...\n",
       "1055449            104/0/0\n",
       "1055450            102/2/0\n",
       "1055451            104/0/0\n",
       "1055452            104/0/0\n",
       "1055453            102/2/0\n",
       "\n",
       "[1055454 rows x 1 columns]"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pur_data_frame = pd.read_csv ('pur.hwe',delimiter=\"\\t\") ## Read the hwe file extrcated from vcf tools \n",
    "keep_col = ['OBS(HOM1/HET/HOM2)'] ## extract only the columns with the genotypes \n",
    "pur_data_frame=pur_data_frame[keep_col] ## save it as a dataframe \n",
    "pur_data_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "4f8d3429-bbd3-4943-89d3-5c6731dca2d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "PUR=pur_data_frame['OBS(HOM1/HET/HOM2)'].str.split('/', expand=True)## split the data frame to get seperate column for each genotype\n",
    "PUR=PUR.rename(columns={PUR.columns[0]:\"PUR_0|0 \"})## rename the columns with the country code and the corresponding genotype symbols \n",
    "PUR=PUR.rename(columns={PUR.columns[1]:\"PUR_0|1 \"})\n",
    "PUR=PUR.rename(columns={PUR.columns[2]:\"PUR_1|1 \"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "6b608281-67f9-490e-ac8b-f1ff08b8618d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PUR_0|0</th>\n",
       "      <th>PUR_0|1</th>\n",
       "      <th>PUR_1|1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055449</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055450</th>\n",
       "      <td>0.980769</td>\n",
       "      <td>0.019231</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055451</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055452</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055453</th>\n",
       "      <td>0.980769</td>\n",
       "      <td>0.019231</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1055454 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         PUR_0|0   PUR_0|1   PUR_1|1 \n",
       "0        1.000000  0.000000       0.0\n",
       "1        1.000000  0.000000       0.0\n",
       "2        1.000000  0.000000       0.0\n",
       "3        1.000000  0.000000       0.0\n",
       "4        1.000000  0.000000       0.0\n",
       "...           ...       ...       ...\n",
       "1055449  1.000000  0.000000       0.0\n",
       "1055450  0.980769  0.019231       0.0\n",
       "1055451  1.000000  0.000000       0.0\n",
       "1055452  1.000000  0.000000       0.0\n",
       "1055453  0.980769  0.019231       0.0\n",
       "\n",
       "[1055454 rows x 3 columns]"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PUR = PUR.astype(int) #convert the date frame date into int \n",
    "PUR.div(104)## divide the genotypes by the number of samples (extracted earlier with allel ) to get the genotype frequencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "9e353124-e2f7-4a7c-84d3-992ff78d9a8b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "107"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## TSI population\n",
    "#### TSI data with allel package ###\n",
    "TSIdata = allel.read_vcf('TSI_SNPs_only.vcf', fields='*') ## Read the VCF file using allel \n",
    "\n",
    "TSIgenotypes=TSIdata['calldata/GT'] ## extract the genotypes array \n",
    "\n",
    "TSIarray=allel.GenotypeArray(TSIgenotypes) ## convert the genotypes data frame into a genotype array \n",
    "\n",
    "TSIarray.n_samples ## count the total number of samples that will be used later to calculate genotypes frequencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "bbb97cab-f6f2-4f9e-a500-841006329d06",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OBS(HOM1/HET/HOM2)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>107/0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>107/0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>107/0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>107/0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>107/0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055449</th>\n",
       "      <td>107/0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055450</th>\n",
       "      <td>106/1/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055451</th>\n",
       "      <td>107/0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055452</th>\n",
       "      <td>107/0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055453</th>\n",
       "      <td>106/1/0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1055454 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        OBS(HOM1/HET/HOM2)\n",
       "0                  107/0/0\n",
       "1                  107/0/0\n",
       "2                  107/0/0\n",
       "3                  107/0/0\n",
       "4                  107/0/0\n",
       "...                    ...\n",
       "1055449            107/0/0\n",
       "1055450            106/1/0\n",
       "1055451            107/0/0\n",
       "1055452            107/0/0\n",
       "1055453            106/1/0\n",
       "\n",
       "[1055454 rows x 1 columns]"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tsi_data_frame = pd.read_csv ('tsi.hwe',delimiter=\"\\t\") ## Read the hwe file extrcated from vcf tools \n",
    "keep_col = ['OBS(HOM1/HET/HOM2)'] ## extract only the columns with the genotypes \n",
    "tsi_data_frame=tsi_data_frame[keep_col] ## save it as a dataframe \n",
    "tsi_data_frame\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "80e0187d-f5b2-4a4b-b2f5-5facdc3c7a48",
   "metadata": {},
   "outputs": [],
   "source": [
    "TSI=tsi_data_frame['OBS(HOM1/HET/HOM2)'].str.split('/', expand=True)## split the data frame to get seperate column for each genotype\n",
    "TSI=TSI.rename(columns={TSI.columns[0]:\"TSI_0|0 \"})## rename the columns with the country code and the corresponding genotype symbols \n",
    "TSI=TSI.rename(columns={TSI.columns[1]:\"TSI_0|1 \"})\n",
    "TSI=TSI.rename(columns={TSI.columns[2]:\"TSI_1|1 \"})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "2b23e84e-04eb-4c45-b5be-992868c7a7ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TSI_0|0</th>\n",
       "      <th>TSI_0|1</th>\n",
       "      <th>TSI_1|1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055449</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055450</th>\n",
       "      <td>0.990654</td>\n",
       "      <td>0.009346</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055451</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055452</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055453</th>\n",
       "      <td>0.990654</td>\n",
       "      <td>0.009346</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1055454 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         TSI_0|0   TSI_0|1   TSI_1|1 \n",
       "0        1.000000  0.000000       0.0\n",
       "1        1.000000  0.000000       0.0\n",
       "2        1.000000  0.000000       0.0\n",
       "3        1.000000  0.000000       0.0\n",
       "4        1.000000  0.000000       0.0\n",
       "...           ...       ...       ...\n",
       "1055449  1.000000  0.000000       0.0\n",
       "1055450  0.990654  0.009346       0.0\n",
       "1055451  1.000000  0.000000       0.0\n",
       "1055452  1.000000  0.000000       0.0\n",
       "1055453  0.990654  0.009346       0.0\n",
       "\n",
       "[1055454 rows x 3 columns]"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TSI = TSI.astype(int) #convert the date frame date into int \n",
    "TSI.div(107)## divide the genotypes by the number of samples (extracted earlier with allel ) to get the genotype frequencies\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c183454a-2e83-44fc-9ee0-eaf5f395db90",
   "metadata": {},
   "source": [
    "## merge all populations frequencies into 1 dataframe "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "1e1b786d-6730-457d-9ccd-d9d26b491eb6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>POS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>16050075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>16050115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>16050213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>16050319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>16050527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055449</th>\n",
       "      <td>51241342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055450</th>\n",
       "      <td>51241386</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055451</th>\n",
       "      <td>51244163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055452</th>\n",
       "      <td>51244205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055453</th>\n",
       "      <td>51244237</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1055454 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              POS\n",
       "0        16050075\n",
       "1        16050115\n",
       "2        16050213\n",
       "3        16050319\n",
       "4        16050527\n",
       "...           ...\n",
       "1055449  51241342\n",
       "1055450  51241386\n",
       "1055451  51244163\n",
       "1055452  51244205\n",
       "1055453  51244237\n",
       "\n",
       "[1055454 rows x 1 columns]"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "##extract the postion from one the data frames \n",
    "POS_data_frame = pd.read_csv ('jpt.hwe',delimiter=\"\\t\") \n",
    "keep_col_POS = ['POS']\n",
    "POS_data_frame=POS_data_frame[keep_col_POS]\n",
    "POS_data_frame\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "0d82b6d9-68f2-4324-8c38-741af4499b60",
   "metadata": {},
   "outputs": [],
   "source": [
    "genotype_frequencies=pd.concat([POS_data_frame,GIH,GWD,JPT,PUR,TSI],axis=1)## concat all the date frames into one data frame withh all the populations and their genotype frequencies "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "09203463-ec55-4119-bd4d-91267e44e3b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "genotype_frequencies.to_csv('Genotype_frequencies_all_population.csv')## convert the data frame into a csv for easy manipulation "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9810c7e-67a7-4f65-82b3-c3c5d3d553af",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
