# Genome Analytics

##  SNP Browser  Software

SNP browser is a fast and accurate application that allows information retrieval of SNPs from Chromosome 22 in conjunction with genotype and allele frequency.This application also allows the user to  perform genetic statistical testing; Tajima's D, Nucleotide diversity and fixation index. The return results are display, plotted, and can be downloaded as a txt file. 

This software was developed by the Genomic Analytics group (4 MSc Bioinformatics students From Queen Mary University) under the guidance of Professor Conrad Bessant and Dr Matteo Fumagalli, in response to the need for an inclusive platform that not only allows the querying and extraction of genomic information but also statistical testing.


## Getting  started - End user

### The following packages will be required  for this application 

Python 3.9.5

### Installing
To run this website from your local machine please download the "website" directory

### Supported enviroments

## Citation and reference

### How shoudl be cited in a document 

#### Licence

#### Contributing







The aim of this project is to create and deployed a piece of software with the following specifications :

The user should be able to retrieve SNP information given either a genomic coordinate (chromosome, start and end), SNP name (rs value), or gene name (or any aliases associated to it).

The application should return the following information for each SNP: name (rs value), genomic position, genotype frequencies, and allele frequency. Frequencies should be provided for each population separately.

If multiple SNPs are returned, the user should be able to select the population(s) and summary statistics of interest, and the application will calculate them and plot their distribution in slidingwindows along the region. The user will also be able to download a text file with the values of summary statistics. At least three summary statistics should be reported, for instance: one index of genetic diversity, one on haplotype diversity (e.g. homozygosity), one test against neutrality (e.g. Tajimaâ€™s D). Additionally, if multiple populations are selected, then population genetic variation (FST value) for each pair of populations should be reported.
